include/dca/linalg/util/cuda_stream.hpp

```diff
@@ -31,10 +31,15 @@ class CudaStream {
   }
 
   CudaStream(const CudaStream& other) = delete;
+  CudaStream& operator=(const CudaStream& other) = delete;
 
   CudaStream(CudaStream&& other) {
     std::swap(stream_, other.stream_);
   }
+  CudaStream& operator=(CudaStream&& other) {
+    std::swap(stream_, other.stream_);
```

Comment:
this is not what I expect operator= to do.  Also I think std::swap only works with cudastream_t because it is just a typedef of a simple type. I'd like to see this be a friend function swap(CudaStream& a, CudaStream& b)

Reply:
Move assignment is very often implemented in terms of swap. Using a moved object is UB, so we can leave it in any state. Leaving it in the previous state  of the object it was assigned to, is the easiest way to ensure proper resources deallocation.
cudaStream_t is just a pointer to the actual resources, so std::swap is well defined. I could make an extra swap method either inside the class or outside, but I did not need it.

Reply:
My knee jerk is more or less like the below blog post.
http://scottmeyers.blogspot.com/2014/06/the-drawbacks-of-implementing-move.html

However I can see this argument really lost out the common usage in the meantime.

Reply:
Of course it is not as performant as a handwritten function with deallocation and setting to nullptr of the other pointer, but all the performance consideration of the blog are not relevant here, as the issue is "non copiability" rather than performance or memory, and the possibly the only relevant cost is the creating and destroying extra Cuda streams.